Sources/SignalProducer.swift

```diff
@@ -229,13 +203,124 @@ public struct SignalProducer<Value, Error: Swift.Error> {
 	///            `Signal` commences. Both the produced `Signal` and an interrupt handle
 	///            of the signal would be passed to the closure.
 	public func startWithSignal(_ setup: (_ signal: Signal<Value, Error>, _ interruptHandle: Disposable) -> Void) {
-		let instance = builder()
-		setup(instance.producedSignal, instance.interruptHandle)
+		return core.startWithSignal(setup)
+	}
+}
+
+internal class SignalProducerCore<Value, Error: Swift.Error> {
```

Comment:
Since there is no generalised existential, we would need a type-erased wrapper if we use protocol.